<a href="https://colab.research.google.com/github/monolixd/Ailearn/blob/main/ActivationAndLoss.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 1. ตัวอย่างการใช้ Activation Functions ใน Dense Layer

In [1]:
import tensorflow as tf
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense

# สร้างโมเดลแบบ Sequential
model = Sequential([
    Dense(64, activation='relu', input_shape=(100,)),  # ReLU
    Dense(32, activation='sigmoid'),                  # Sigmoid
    Dense(10, activation='softmax')                   # Softmax
])

# แสดงสรุปโมเดล
model.summary()

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense (Dense)                        │ (None, 64)                  │           6,464 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 32)                  │           2,080 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_2 (Dense)                      │ (None, 10)                  │             330 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 8,874 (34.66 KB)

 Trainable params: 8,874 (34.66 KB)

 Non-trainable params: 0 (0.00 B)

Output นี้แสดง Summary ของโครงสร้างโมเดล Neural Network โดยอธิบายลักษณะของแต่ละเลเยอร์ในโมเดล และจำนวนพารามิเตอร์ที่ใช้ในการเทรนโมเดล โดยมีรายละเอียดดังนี้:

Column 1: Layer (type)

แสดงชื่อเลเยอร์และประเภทเลเยอร์ที่ใช้ (ในที่นี้ใช้ Dense layers ทั้งหมด ซึ่งเป็น Fully Connected Layers)
Column 2: Output Shape

แสดงขนาดของ Output ที่ได้จากแต่ละเลเยอร์
(None, 64) หมายถึง ข้อมูลมี 64 units/output neurons (ค่า None คือ Batch Size ที่สามารถปรับเปลี่ยนได้)
Column 3: Param #

แสดงจำนวนพารามิเตอร์ (Parameters) ที่ใช้ในแต่ละเลเยอร์
จำนวนนี้คิดจากการคูณระหว่าง input features และ output neurons รวมกับ bias parameters
ตัวอย่าง:
Dense (64 units): (input_features + 1) * output_units = (100 + 1) * 64 = 6,464
Dense_1 (32 units): (64 + 1) * 32 = 2,080
Dense_2 (10 units): (32 + 1) * 10 = 330
Summary ของพารามิเตอร์

Total params: จำนวนพารามิเตอร์ทั้งหมดในโมเดล = 8,874
Trainable params: จำนวนพารามิเตอร์ที่สามารถปรับปรุงได้ผ่านการเทรน = 8,874
Non-trainable params: พารามิเตอร์ที่ไม่ถูกปรับในระหว่างการเทรน (ไม่มีในโมเดลนี้ = 0)

## 2. การใช้ Activation Functions แบบกำหนดเอง

In [ ]:
from tensorflow.keras import activations

# ใช้ Activation Function ในโค้ดแบบกำหนดเอง
relu_output = activations.relu([-3.0, 0.0, 3.0])  # ค่า Output: [0.0, 0.0, 3.0]
sigmoid_output = activations.sigmoid([-3.0, 0.0, 3.0])  # ค่า Output: [0.047, 0.5, 0.952]
print("ReLU Output:", relu_output.numpy())
print("Sigmoid Output:", sigmoid_output.numpy())

activations.relu: แปลงค่าที่น้อยกว่า 0 ให้เป็น 0<br>
activations.sigmoid: แปลงค่าทั้งหมดให้อยู่ในช่วง (0, 1)